python-avd/pyavd/_schema/models/avd_model.py

```diff
@@ -0,0 +1,341 @@
+# Copyright (c) 2023-2024 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+from __future__ import annotations
+
+from collections.abc import Mapping
+from typing import TYPE_CHECKING, Any, ClassVar, Literal
+
+from pyavd._schema.coerce_type import coerce_type
+from pyavd._utils import Undefined
+
+from .avd_base import AvdBase
+from .avd_indexed_list import AvdIndexedList
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+    from .type_vars import T_AvdModel
+
+
+class AvdModel(AvdBase):
+    _is_avd_model = True
+    _allow_other_keys: bool = False
+    _fields: ClassVar[dict[str, dict]]
+    _required_fields: ClassVar[tuple[str, ...]]
+    _field_to_key_map: ClassVar[dict[str, str]]
+    _key_to_field_map: ClassVar[dict[str, str]]
+
+    @classmethod
+    def _from_dict(cls: type[T_AvdModel], data: Mapping) -> T_AvdModel:
+        """Returns a new instance loaded with the data from the given dict."""
+        if not isinstance(data, Mapping):
+            msg = f"Expecting 'data' as a 'Mapping' when loading data into '{cls.__name__}'. Got '{type(data)}"
+            raise TypeError(msg)
+
+        has_custom_data = "_custom_data" in cls._fields
+        cls_args = {}
+
+        for key in data:
+            if has_custom_data and str(key).startswith("_"):
+                cls_args.setdefault("_custom_data", {})[key] = data[key]
+                continue
+
+            if not (field := cls._get_field_name(key)):
+                if cls._allow_other_keys:
+                    # Ignore unknown keys.
+                    continue
+                msg = f"Invalid key '{key}'. Not available on '{cls.__name__}'."
+                raise KeyError(msg, has_custom_data)
+
+            field_info = cls._fields[field]
+            field_type = field_info["type"]
+
+            value = coerce_type(data[key], field_type, list_items_type=field_info.get("items"))
+
+            # Raise for wrong type ignoring None values - we expect the validation to have sorted out required fields.
+            if value is not None and not isinstance(value, field_type):
+                msg = f"Invalid type '{type(value)}. Expected '{field_type}'. Value '{value}"
+                raise TypeError(msg)
```

Comment:
This should never happen as long as the schema tooling ran first. But yes, it is a safeguard in case of validation running with warnings and we cannot coerce.